rust/src/io/exec/planner.rs

```diff
@@ -95,6 +95,12 @@ impl Planner {
             UnaryOperator::Not | UnaryOperator::PGBitwiseNot => {
                 Expr::Not(Box::new(self.parse_sql_expr(expr)?))
             }
+
+            UnaryOperator::Minus => {
+                let num_literal = format!("-{}", expr.to_string());
+                return self.number(&num_literal);
```

Comment:
Try this:
```suggestion
                lit(-1) * self.parse_sql_expr(expr)?
```

Reply:
Hi Will,
I tried a couple combinations, most of them couldn't pass the basic cases
Here are the examples
```rust
lit(-1) * self.parse_sql_expr(expr)?
```
```python
print(dataset.scanner(filter="x = -1").to_table())
```
pyarrow.lib.ArrowInvalid: C Data interface error: Io error: Execution error: LanceError(IO): Internal error: Cannot evaluate binary expression Multiply with types Int32 and Int64. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker

***Comment*** - This seems fixable

Converting -1 to Int64 should resolve this ideally
```rust
lit(-1 as i64) * self.parse_sql_expr(expr)?
```
pyarrow.lib.ArrowInvalid: C Data interface error: Io error: Execution error: LanceError(IO): Internal error: Cannot evaluate binary expression Eq with types Float64 and Int64. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker

***Comment*** - In this case it parsed the -1 as an expr correctly but failed in comparison.  If I change the i64 for f64 then the error will be like of previous case where binary operation mulitply cannot be performed for f64 and i64.

The inability of execution engine to deal with cross types operations is making even trivial cases throw errors.

We could either reduce the scope of support, where only basic constants (1, 1.0, -1.0 ) are supported and complex expressions like -1*1 are not. 
Or
We could force convert every numeric value to f64, which again doesn't seem right.

What are your thoughts ?

Reply:
I tried that before arriving at first solution, the error thrown is this

ValueError: LanceError(IO): Expression '(- Int64(1))' is not supported as filter in lance

Apparently Expr::Negative adds a space between symbol and operand which can't be executed as a filter rather than negating the expression

from datafusion
```rust
Expr::Negative(expr) => {
    let expr = create_name(expr)?;
    Ok(format!("(- {expr})"))
}
```

What we essentially want is for Expr::Value i.e constants, - operand used in unary expr should multiply the value by -1 but keep whatever type it was f64/i64 and keep it part of the rest of the expr as same.

```rust
  match expr {
      SQLExpr::Value(val) => {
          let num_literal = format!("-{}", val.to_string());
          return Ok(self.number(&num_literal)?)
      }
      _ => Expr::Negative(Box::new(self.parse_sql_expr(expr)?))
  }
```
This implementation works with basic cases. Is able to parse cases like "x = 1 + -1" but throws still throws error like

pyarrow.lib.ArrowInvalid: C Data interface error: Io error: Execution error: LanceError(IO): Internal error: Cannot evaluate binary expression Eq with types Float64 and Int64. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker

This is very strange behaviour, there is no float in this expr, as -1 is parsed to i64.

Reply:
> I tried that before arriving at first solution, the error thrown is this

It sounds like you handled the translation from SQL to Expr, but not Expr to PhysicalExpr. I think [NegativeExpr](https://docs.rs/datafusion/26.0.0/datafusion/physical_plan/expressions/struct.NegativeExpr.html) is what you will map it to. Did you try that?

Reply:
Added this in create_physical_expr function
```rust
  Expr::Negative(expr) => Arc::new(NegativeExpr::new(
      self.create_physical_expr(expr.as_ref())?
  )),
```
Still getting the same error for "x = -1" case.

pyarrow.lib.ArrowInvalid: C Data interface error: Io error: Execution error: LanceError(IO): Internal error: Cannot evaluate binary expression Eq with types Float64 and Int64. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker

After printing the expr in create_physical_expr function, where essentially binary expr "eq" was converted to physical expr I found some strange things
expr = " x = 1" => x = Float64(1) i.e 1 was converted to Float64 for it to work. 
Expr::Negative was converting "x = -1" to x = (- Int64(1)) hence the type mismatch error.

***If*** we parse all numbers as f64, we can avoid this error for all numeric expressions.

What do you think ?

Reply:
I don't think parsing integers as floats is a good idea. f64 can't represent all i64, and we want users to be able to represent them all. Think of the case where there is an id column that is i64. We would want the float conversion rounding the unique id, causing a query to get the wrong record.

The filter works fine if you write the expression as a float (with a trailing decimal):

```sql
x = -1.0
```

Looking at what DataFusion does, I wonder if we just improve our parsing logic we will get that automatic conversion you are seeing for negative numbers too. See here:

https://github.com/apache/arrow-datafusion/blob/9b419b19a66bdd35e9e5c0bca259786f8f3c3965/datafusion/sql/src/expr/unary_op.rs#L39-L50

Reply:
Hi Will,
Thank you for your suggestion and digging. I updated our SQLExpr Unary Op parsing like you said and SQL expression parsing did work fine.
After some more digging I figured that type conversion was happening to Float64 due to type of our column x in table was double. This is when we resolve logical expr with schema to produce resolved expr. 
In llance::datafusion::logical_expr::resolove_expr this activity takes place. 
Up until now it only supported binary expressions with columns on one side and literals on other, hence type coercion was/is successfull for trivial expressions.

I've added new function coerce_expr to convert a binary expr to column type. With this all the cases below work as expected
```python
tab = pa.table({"x": [-1, 1, 0, 1.0]})
dataset = lance.write_dataset(tab, "lance_values")

dataset.scanner(filter="x = 1").to_table()
dataset.scanner(filter="x = -1").to_table()
dataset.scanner(filter="x > -1").to_table()
dataset.scanner(filter="x = 1 * 1.0").to_table()
dataset.scanner(filter="x = 2 - 2 + -1").to_table()
```

Please let me know your thoughts over this.

Reply:
I think I messed my rebase. My changes are only in the last commit
https://github.com/lancedb/lance/pull/983/commits/81836ea867b2b0e2bf8f0b3336adedf699834347
I'll fix that.